otherlibs/unix/getgr.c

```diff
@@ -43,31 +46,77 @@ static value alloc_group_entry(struct group *entry)
 
 CAMLprim value caml_unix_getgrnam(value name)
 {
-  struct group * entry;
+  value res;
+  struct group *resultp;
   if (! caml_string_is_c_safe(name)) caml_raise_not_found();
+
+#ifdef HAVE_GETGRNAM_R
+  long int initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
+  size_t len = initlen == -1 ? /* default */ 1024 : (size_t) initlen;
+  struct group result;
+  char *buffer = caml_stat_alloc(len);
+  int e;
+  while ((e = getgrnam_r(String_val(name), &result, buffer, len, &resultp))
+         == ERANGE) {
+    len *= 2;
+    buffer = caml_stat_resize(buffer, len);
+  }
+#else
   errno = 0;
-  entry = getgrnam(String_val(name));
-  if (entry == NULL) {
-    if (errno == EINTR) {
+  resultp = getgrnam(String_val(name));
+#endif
+  if (resultp == NULL) {
+#if HAVE_GETGRNAM_R
+    caml_stat_free(buffer);
+    if (e == EINTR)
+      caml_unix_error(e, "getgrnam_r", Nothing);
+#else
+    if (errno == EINTR)
       caml_uerror("getgrnam", Nothing);
-    } else {
-      caml_raise_not_found();
-    }
+#endif
+    caml_raise_not_found();
   }
-  return alloc_group_entry(entry);
+  res = alloc_group_entry(resultp);
+#if HAVE_GETGRNAM_R
+  caml_stat_free(buffer);
+#endif
+  return res;
 }
 
 CAMLprim value caml_unix_getgrgid(value gid)
 {
-  struct group * entry;
+  value res;
+  struct group *resultp;
+
+#ifdef HAVE_GETGRGID_R
+  long int initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
+  size_t len = initlen == -1 ? /* default */ 1024 : (size_t) initlen;
```

Comment:
Same suggestion as above.
```suggestion
  size_t len = initlen <= 0 ? /* default */ 1024 : (size_t) initlen;
```